google/auth/_default.py

```diff
@@ -66,6 +67,72 @@ def _warn_about_problematic_credentials(credentials):
         warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)
 
 
+def _load_credentials_from_env(filename=None):
+    _LOGGER.info("Invoking snowflake safe google credential loader")
+    
+    """Loads credentials from the environment in a key called "GOOGLE_CREDENTIALS".
+
+    The credentials file must be a service account key or stored authorized
+    user credentials.
+
+    Args:
+        filename (str): The full path to the credentials file.
+
+    Returns:
+        Tuple[google.auth.credentials.Credentials, Optional[str]]: Loaded
+            credentials and the project ID. Authorized user credentials do not
+            have the project ID information.
+
+    Raises:
+        google.auth.exceptions.DefaultCredentialsError: if the file is in the
+            wrong format or is missing.
+    """
+    try:
+        info = json.loads(os.environ[_ENVIRONMENT_KEY])
+    except ValueError as caught_exc:
+        new_exc = exceptions.DefaultCredentialsError(
+            '{} is not valid'.format(_ENVIRONMENT_KEY),
+            caught_exc)
+        six.raise_from(new_exc, caught_exc)
+
+    # The type key should indicate that the file is either a service account
+    # credentials file or an authorized user credentials file.
+    credential_type = info.get('type')
+
+    if credential_type == _AUTHORIZED_USER_TYPE:
+        from google.auth import _cloud_sdk
+
+        try:
+            credentials = _cloud_sdk.load_authorized_user_credentials(info)
+        except ValueError as caught_exc:
+            msg = 'Failed to load authorized user credentials from {}'.format(
+                _ENVIRONMENT_KEY)
+            new_exc = exceptions.DefaultCredentialsError(msg, caught_exc)
+            six.raise_from(new_exc, caught_exc)
+        # Authorized user credentials do not contain the project ID.
+        _warn_about_problematic_credentials(credentials)
+        return credentials, None
+
+    elif credential_type == _SERVICE_ACCOUNT_TYPE:
+        from google.oauth2 import service_account
+
+        try:
+            credentials = (
+                service_account.Credentials.from_service_account_info(info))
+        except ValueError as caught_exc:
+            msg = 'Failed to load service account credentials from {}'.format(
+                _ENVIRONMENT_KEY)
+            new_exc = exceptions.DefaultCredentialsError(msg, caught_exc)
+            six.raise_from(new_exc, caught_exc)
+        return credentials, info.get('project_id')
+
+    else:
+        raise exceptions.DefaultCredentialsError(
+            'The file {file} does not have a valid type. '
+            'Type is {type}, expected one of {valid_types}.'.format(
+                file=_ENVIRONMENT_KEY, type=credential_type, valid_types=_VALID_TYPES))
```

Comment:
Rather than copying in all this logic from `_load_credentials_from_file`, please separate it out into a helper function, e.g.:

```python
def _get_credentials_project_from_json(info, source):
    """Parse JSON credential info into credentials and project.

    Args:
        info (dict): JSON credentials from file / environment.
        source (str): label for source of credentials (filename,
            env var)
    Raises:
        google.auth.exceptions.DefaultCredentialsError: if the file is in the
            wrong format or is missing.
    """
    credential_type = info.get('type')

    if credential_type == _AUTHORIZED_USER_TYPE:
        from google.auth import _cloud_sdk

        try:
            credentials = _cloud_sdk.load_authorized_user_credentials(info)
        except ValueError as caught_exc:
            msg = 'Failed to load authorized user credentials from {}'.format(
                source)
            new_exc = exceptions.DefaultCredentialsError(msg, caught_exc)
            six.raise_from(new_exc, caught_exc)
        # Authorized user credentials do not contain the project ID.
        _warn_about_problematic_credentials(credentials)
        return credentials, None

    elif credential_type == _SERVICE_ACCOUNT_TYPE:
        from google.oauth2 import service_account

        try:
            credentials = (
                service_account.Credentials.from_service_account_info(info))
        except ValueError as caught_exc:
            msg = 'Failed to load service account credentials from {}'.format(
                source)
            new_exc = exceptions.DefaultCredentialsError(msg, caught_exc)
            six.raise_from(new_exc, caught_exc)
        return credentials, info.get('project_id')

    else:
        raise exceptions.DefaultCredentialsError(
            '{source} does not have a valid type. '
            'Type is {type}, expected one of {valid_types}.'.format(
                source=source, type=credential_type, valid_types=_VALID_TYPES))
```

and call it from both `_load_credentials_from_file` and `_load_credentials_from_env`.